### PR TITLE
chore(node): mark object values as undefine-able

### PR DESCRIPTION
**Pull request by SimonSchick**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: Does not apply
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

NodeJS typings have been a little inconsistent with their use of making objects with index signatures have ` | undefined` for their properties.

[Which I have contributed to...](https://github.com/DefinitelyTyped/DefinitelyTyped/commit/2b7def6f58415d444b84aca6f6cb1c4b216bd403#r38450075)

This change introduces 2 utility dictionary types that make things more consistent.

Pro:
- More type safety
- Consistency

Contra:
- This is breaking (however breaking changes have been allowed in the past if they contribute to type safety or types were outright wrong).
- These types should probably be provided by TS itself as utilities (similar to `ReadOnlyArray` and friends).
- There is an ongoing debate about this (see link above) and https://github.com/Microsoft/TypeScript/issues/13778

### CODE REVIEW

File: types/node/globals.d.ts [hunk 1]
```diff
@@ -1066,15 +1065,11 @@ declare namespace NodeJS {
...
-    interface NodeRequireCache {
-        [path: string]: NodeModule;
-    }
-
     interface Require {
         /* tslint:disable-next-line:callable-types */
         (id: string): any;
         resolve: RequireResolve;
-        cache: NodeRequireCache;
+        cache: Dict<NodeModule>;
```
Comment (msheakoski) [inline types/node/globals.d.ts:1072]: This introduces "undefined" as a possible value for cache, where before it must be an object. I am not yet sure how Node.js handles this internally, but just wanted to mark a potential difference from the previous code.

Comment (msheakoski) [reply types/node/globals.d.ts:1072]: Oops! Nevermind. I misunderstood. Please ignore!

**Comment by msheakoski** [comment]
Thank you for making this PR, @SimonSchick. Even though you're on team no-undefined 😉, I think that this is a good move since it results in more consistency!

**Comment by SimonSchick** [comment]
@msheakoski I think you got me wrong
I'd personally *prefer* if TS assumed undefined for index sigs by default as it's more aligned with strict null checks!

I've had too many bugs where undefined in an index singature cost me valuable time :P

**Comment by msheakoski** [comment]
@SimonSchick My apologies for the misunderstanding! I guess we're on the same team after all! 🤗

**Review by Flarna — APPROVED** [review]

**Comment by nstepien** [comment]
```
node_modules/@types/node/globals.d.ts:1075:15 - error TS2430: Interface 'Require' incorrectly extends interface 'RequireFunction'.
  Types of property 'cache' are incompatible.
    Type 'Dict<NodeModule>' is not assignable to type '{ [id: string]: NodeModule; }'.
      Index signatures are incompatible.
        Type 'NodeModule | undefined' is not assignable to type 'NodeModule'.
          Type 'undefined' is not assignable to type 'NodeModule'.

1075     interface Require {
                   ~~~~~~~
```
Looks like there's a type conflict with `@types/webpack-env`.

**Comment by gitter-me** [comment]
For a unknown reason, the change in line [types/node/globals.d.ts:673](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/43931/files?diff=unified&w=1#diff-a2f9a5377787f7084c7f52b20c0108cfR673) conflicts with Create-React-App [react-scripts/lib/react-app.d.ts:6](https://github.com/facebook/create-react-app/blob/master/packages/react-scripts/lib/react-app.d.ts#L6).
The error message is Duplicate identifier 'ProcessEnv' 

I use Typescript 3.8.3, @types/node 13.13.0 and react-scripts 3.4.1

I'm not sure if this error is only on my machine, and if it is a general issue, where it is better placed (types/node or create-react-app)?

**Comment by peterblazejewicz** [comment]
@gitter-me
see #44003

**Comment by OliverJAsh** [comment]
Now TS finally has `noUncheckedIndexedAccess`, perhaps we should remove `undefined` from the dictionary. Then it's up to users what behaviour they want—they can enable/disable `noUncheckedIndexedAccess` to their liking.

One problem with adding `undefined` to the index signature is that you may end up with `undefined` when using an object's values in other ways, outside of just indexing:

```ts
declare const dict: { [key: string]: number | undefined };

const value = dict['foo']; // number | undefined ✅

Object.values(dict).map((value) => {
    value; // number | undefined; ❌
});
```

This doesn't occur with `noUncheckedIndexedAccess`:

```ts
declare const dict: { [key: string]: number };

const value = dict['foo']; // number | undefined ✅

Object.values(dict).map((value) => {
    value; // number; ✅
});
```

WDYT? @SimonSchick @Flarna

**Pull request merged**